File: tests/tokenizer/test_tokenizer.cpp
#include "../src/tokenizer/tokenizer.hpp"
#include "../cmdline.hpp"


int main(int argc, char *argv[])
{
    std::string tokenizer_path = "../tests/tokenizer.txt";
    cmdline::parser a;
    a.add<std::string>("tokenizer_path", 't', "tokenizer path", true);
    a.add<std::string>("system_prompt", 0, "system prompt", false, "你是一个翻译助手，无论我说什么，都翻译成英文");
    a.add<std::string>("text", 0, "text", true);
    a.parse_check(argc, argv);
    tokenizer_path = a.get<std::string>("tokenizer_path");
    std::string text = a.get<std::string>("text");
    std::string system_prompt = a.get<std::string>("system_prompt");

    std::unique_ptr<MNN::Transformer::Tokenizer> tokenizer(MNN::Transformer::Tokenizer::createTokenizer(tokenizer_path));

    std::string chat_template = "<|im_start|>system\n%s<|im_end|>\n<|im_start|>user\n%s<|im_end|>\n<|im_start|>assistant\n";
    
    char prompt[1024];
    sprintf(prompt, chat_template.c_str(), system_prompt.c_str(), text.c_str());
  

    printf("prompt: %s\n", prompt);
    auto ids = tokenizer->encode(prompt);
    printf("ids size: %ld\n", ids.size());
    for (auto id : ids)
    {
        std::cout << id << ", ";
    }
    std::cout << std::endl;

    std::string _text;
    for (auto id : ids)
    {
        _text += tokenizer->decode(id);
    }
    std::cout << "text: \n" << _text << std::endl;

    return 0;
}

File: tests/tokenizer/test_tokenizer.py
import os
import json
import glob
import base64
import argparse

from transformers import AutoTokenizer

def main():
    parser = argparse.ArgumentParser(description='llm_exporter', formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('--tokenizer_path', type=str, help='tokenizer path')
    parser.add_argument('--system_prompt', type=str, help='system prompt', default="你是一个翻译助手，无论我说什么，都翻译成英文")
    parser.add_argument('--text', type=str, help='')
    args = parser.parse_args()


    try:
        tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_path, trust_remote_code=True, use_fast=False)
    except:
        tokenizer = None
    if None == tokenizer:
        try:
            tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_path, trust_remote_code=True)
        except:
            tokenizer = None
    if None == tokenizer:
        print("Default load tokenizer failed for ", args.tokenizer_path)
    
    messages = [
            {"role": "system", "content": args.system_prompt},
            {"role": "user", "content": args.text},
        ]
    
    text = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=True
        )
    print(text)
    
    ids = tokenizer.encode(text)
    print("ids size: ", len(ids))
    print("ids: ", ids)
    
    text = tokenizer.decode(ids)
    print("text: \n", text)
    
        # text = tokenizer.decode([0, 994, 1322, 2])
        # print("text: ", text)

if __name__ == '__main__':
    main()